modules/ROOT/pages/_partials/sync/server-creating-a-datasource-schema-mapping.adoc

```diff
@@ -0,0 +1,137 @@
+[id="server-creating-a-datasource-schema-mapping-{context}"]
+= Creating a data source schema mapping
```

Comment:
I am not really sure what the title means here. Also from the code example, I am not really sure where does it shows the `mapping` part. Does this mean how you can map an existing database schema to graphql?

Reply:
This idea here is the following:

I have an existing database and in my current Mobile App, I run a sql statement similar to this:
---
Select employee_name
From Employee
Where Employee.Category in (1,2)
---
The goal of this work is this - I want to be able to create a schema that maps to an existing database schema, something like:
---
// Stub that reflects data source mapping
const Employee = [{
    employee_id: 1,
    employee_name:"Eddie",
    Category: {
        category_id: 1,
        employee_type: "Manager",
        security_clearance: 1
    }
},
{
    employee_id: 2,
    employee_name:"Max",
    Category: {
        category_id: 1,
        employee_type: "Associate",
        security_clearance: 2
    }
}]
---

Reply:
ok, in this case then the stub data isn't really valid. If I understand correctly, the schema here has 2 tables, one is `Employee`, one is `Category`. So if you run this query `Select employee_name From Employee Where Employee.Category in (1,2)`, you will get:

```
const Employee = [{
employee_id: 1,
employee_name:"Eddie",
Category: 1
},
{
employee_id: 2,
employee_name:"Max",
Category: 2
}]
```

and in order to get the category data, you need to run another query `Select category_id, employee_type, security_clearance WHERE category_id in (1,2)`, and this will return a separate list that only contains the category information. You can use join statement to combine them into one query, but neither of them will give you back this nested data format.

So I think the right setup should be:

data source schema:

```
[.Employee]
|===
| employee_id, `ID`
| employee_name, `String`
| category_id, `ID`
|===

[.Category]
|===
| category_id, `ID`
| employee_type, `String`
| security_clearance, `ID`
|===
```

graphql def for employee:

```
type Employee {
      employee_id: ID!,
      employee_name: String!
      category: Category
  }
```

and you have 2 lists of data that could be returned from sql:

```
const employees = [{employee_id: 1, employee_name: "joe", cateory_id: 1 }, {employee_id: 2, employee_name: "john", cateory_id: 2}];
const categories = [{category_id: 1, employee_type:"Manager", security_clearance: 1}, {category_id: 2, employee_type:"Associate", security_clearance: 2}];
```

Given those 2 list of data, to resolver map should be:

```
const resolvers = {
  Query: {
    listEmployees: (obj, args, context, info) => {
      return employees;
    }
  },
  Employee: {
    category(employee) {
        return filter(categories, {category_id: employee.cateory_id}); 
    }
  }
}
``` 

See https://www.apollographql.com/docs/graphql-tools/resolvers#resolver-map.

Reply:
@wei-lee - that is outstanding feedback, fair play! :+1: X :100:

Reply:
Ok, from a PostgreSQL pov - I'm hoping to connect into a schema like:
![Screenshot from 2019-05-10 11-15-44](https://user-images.githubusercontent.com/19431386/57520347-55a69a00-7315-11e9-9b21-5ea425be5baa.png)